File: python/Trees/maximum_depth_binary_tree.py
from typing import Optional
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution2:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        """
        Find the maximum depth of a binary tree using BFS (level order traversal).
        Approach:
        - Use a queue to traverse the tree level by level.
        - Increment level counter for each level traversed.
        Time Complexity: O(n)
        Space Complexity: O(w) (w = max width of tree)
        """
        if not root:
            return 0
        level = 0
        q = deque([root])
        while q:
            for i in range(len(q)):
                node = q.popleft()
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            level += 1
        return level

class Solution3:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        """
        Find the maximum depth of a binary tree using iterative DFS approach.
        Strategy:
        1. Use a stack to simulate DFS traversal
        2. Keep track of current depth for each node
        3. Update maximum depth when reaching leaf nodes
        Time Complexity: O(n) where n is the number of nodes
        Space Complexity: O(h) where h is the height of the tree (O(n) in worst case for skewed tree)
        """
        if not root:
            return 0
        stack = [(root, 1)]
        res = 0
        while stack:
            node, depth = stack.pop()
            if node:
                res = max(res, depth)
                stack.append((node.right, depth + 1))  # type: ignore
                stack.append((node.left, depth + 1))   # type: ignore
        return res

# -------------------
# Helper Functions for Testing
# -------------------
def build_tree_from_list(lst):
    if not lst:
        return None
    nodes = [TreeNode(val) if val is not None else None for val in lst]
    kids = nodes[::-1]
    root = kids.pop()
    for node in nodes:
        if node:
            if kids: node.left = kids.pop()
            if kids: node.right = kids.pop()
    return root

# -------------------
# Test Cases
# -------------------
def test_max_depth():
    s = Solution()
    # Test case 1: Example
    root = build_tree_from_list([3,9,20,None,None,15,7])
    assert s.maxDepth(root) == 3
    # Test case 2: Single node
    root = build_tree_from_list([1])
    assert s.maxDepth(root) == 1
    # Test case 3: Empty tree
    root = build_tree_from_list([])
    assert s.maxDepth(root) == 0
    # Test case 4: Left-skewed tree
    root = build_tree_from_list([1,2,None,3,None,4])
    assert s.maxDepth(root) == 4
    # Test case 5: Right-skewed tree
    root = build_tree_from_list([1,None,2,None,3,None,4])
    assert s.maxDepth(root) == 4
    print("All test cases passed!")

if __name__ == "__main__":
    test_max_depth()

File: C++/Trees/maximum_depth_binary_tree.cpp
#include <bits/stdc++.h>  // includes all standard C++ libraries
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    /**
     * Find maximum depth of a binary tree using iterative DFS.
     *
     * Strategy:
     * - Use a stack to perform depth-first traversal.
     * - At each step, track the depth of the current node.
     * - Update the maximum depth whenever visiting a node.
     *
     * Time Complexity: O(n) — visit each node once.
     * Space Complexity: O(h) — maximum stack size equals height of tree.
     */
    int maxDepth(TreeNode* root) {
    if (!root) return 0;

    stack<pair<TreeNode*, int>> stk;
    stk.push({root, 1});
    int res = 0;

    while (!stk.empty()) {
        pair<TreeNode*, int> top = stk.top(); // C++11 style
        stk.pop();

        TreeNode* node = top.first;
        int depth = top.second;

        if (node) {
            res = max(res, depth);

            if (node->right) stk.push({node->right, depth + 1});
            if (node->left) stk.push({node->left, depth + 1});
        }
    }

    return res;

    }
};

/*
-------------------- LINE-BY-LINE SYNTAX EXPLANATION --------------------

#include <bits/stdc++.h>
    ✅ Includes all standard C++ headers (like <vector>, <stack>, <iostream> etc.)

using namespace std;
    ✅ Allows you to use standard library names without writing std:: before them.

struct TreeNode {
    ✅ Defines a structure to represent a binary tree node.

int val;
    ✅ Holds the integer value of the current node.

TreeNode *left, *right;
    ✅ Pointers to the left and right child nodes.

TreeNode() : val(0), left(nullptr), right(nullptr) {}
    ✅ Default constructor: value = 0, children are null.

TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    ✅ Constructor with only value.

TreeNode(int x, TreeNode *left, TreeNode *right)
    ✅ Constructor with value and pointers to left and right children.

class Solution {
    ✅ Defines the Solution class.

public:
    ✅ Public methods that can be accessed outside the class.

int maxDepth(TreeNode* root)
    ✅ Function to compute the maximum depth of the binary tree.

if (!root) return 0;
    ✅ If root is null, tree is empty, depth is 0.

stack<pair<TreeNode*, int>> stk;
    ✅ Stack to store pairs of (node, depth).

stk.push({root, 1});
    ✅ Push the root node with depth 1.

int res = 0;
    ✅ Variable to store maximum depth found.

while (!stk.empty())
    ✅ Loop until the stack becomes empty.

auto [node, depth] = stk.top();
    ✅ Structured binding: get the top element (node & depth).

stk.pop();
    ✅ Remove the top element.

if (node)
    ✅ If the node is not null.

res = max(res, depth);
    ✅ Update the maximum depth if current depth is larger.

if (node->right) stk.push({node->right, depth + 1});
if (node->left) stk.push({node->left, depth + 1});
    ✅ Push right and left children with updated depth.

return res;
    ✅ Return the maximum depth of the tree.

*/
